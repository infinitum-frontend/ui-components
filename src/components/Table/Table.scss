.inf-table {
  border-collapse: separate;
  border-spacing: 0;
  width: 100%;
  font-family: var(--inf-font-family-base);
  --table-border-radius: 0;
  border-radius: var(--table-border-radius);

  thead {
    th {
      background-color: var(--inf-color-background-tertiary);
      text-align: left;
      color: var(--inf-color-text-secondary);
      font: var(--inf-font-body-1);
      padding: var(--inf-space-medium);
      position: relative;
      border-top: 1px solid var(--inf-color-border-primary);

      &:hover .inf-table-header__resizer {
        opacity: 1;
      }

      &:first-child {
        border-top-left-radius: var(--table-border-radius);
        border-left: 1px solid var(--inf-color-border-primary);
      }

      &:last-child {
        border-top-right-radius: var(--table-border-radius);
        border-right: 1px solid var(--inf-color-border-primary);
      }
    }
  }

  tbody {
    tr {
      font: var(--inf-font-body-2);

      @include hover {
        background-color: var(--inf-color-background-primary-hover);
      }

      &.inf-table__row--interactive {
        cursor: pointer;
      }

      &.inf-table__row--selected {
        background-color: var(
          --inf-color-background-primary-hover
        ); // TODO: уточнить цвет
      }

      &:not(:last-child) {
        border-bottom: 1px solid var(--inf-color-border-primary);
      }
    }

    tr:last-child td:first-child {
      border-bottom-left-radius: var(--table-border-radius);
    }

    tr:last-child td {
      border-bottom: 1px solid var(--inf-color-border-primary);
    }

    tr:last-child td:last-child {
      border-bottom-right-radius: var(--table-border-radius);
    }

    td {
      padding: var(--inf-space-medium);
      font: var(--inf-font-body-1);
      color: var(--inf-color-text-primary);
      border-top: 1px solid var(--inf-color-border-primary);
    }

    td:first-child {
      border-left: 1px solid var(--inf-color-border-primary);
    }

    td:last-child {
      border-right: 1px solid var(--inf-color-border-primary);
    }
  }

  &--border-radius-xsmall {
    --table-border-radius: var(--inf-border-radius-xsmall);
  }

  &--border-radius-small {
    --table-border-radius: var(--inf-border-radius-small);
  }

  &--border-radius-medium {
    --table-border-radius: var(--inf-border-radius-medium);
  }

  &--border-radius-large {
    --table-border-radius: var(--inf-border-radius-large);
  }

  &--borderless {
    --table-border-radius: 0;

    thead th {
      border-top: none;

      &:first-child,
      &:last-child {
        border-left: none;
        border-right: none;
      }
    }

    tbody {
      tr td {
        &:first-child,
        &:last-child {
          border-left: none;
          border-right: none;
        }
      }

      tr:last-child td {
        border-bottom: none;
      }

      tr:first-child td {
        border-top: none;
      }
    }
  }

  $vertical-align-values: (
    'baseline',
    'bottom',
    'middle',
    'sub',
    'super',
    'text-bottom',
    'text-top',
    'top'
  );

  @mixin vertical-align($vertical-align-value) {
    &--vertical-align-#{$vertical-align-value} td {
      vertical-align: #{$vertical-align-value};
    }
  }

  @each $vertical-align-value in $vertical-align-values {
    @include vertical-align($vertical-align-value);
  }
}
