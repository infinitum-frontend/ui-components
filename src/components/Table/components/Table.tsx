import {
  CSSProperties,
  ReactElement,
  TableHTMLAttributes,
  useState
} from 'react'
import {
  ColumnDef,
  getCoreRowModel,
  getFilteredRowModel,
  useReactTable,
  SortingState,
  getSortedRowModel
} from '@tanstack/react-table'
import cn from 'classnames'
import TableHeader from 'Components/Table/components/Header'
import TableBody from 'Components/Table/components/Body'
import '../index.scss'

interface BaseRow {
  className?: string
  style?: CSSProperties
}

export type Row<T extends Record<any, any>> = BaseRow & T
export interface TableProps extends TableHTMLAttributes<HTMLTableElement> {
  /** Массив с данными для построения шапки таблицы */
  columns: Array<ColumnDef<any, any>>
  /** Массив с данными для построения тела таблицы */
  rows: Array<Row<any>>
  /** Максимальное количество отображаемых элементов */
  maxLength?: number
  /** Включение сортировки по столбцам */
  withSorting?: boolean
  sortingState?: SortingState
  // /** Отображение чекбоксов в 1 колонке */
  // withRowSelection?: boolean
  // withFiltering?: boolean
  // /** Событие изменения чекбоксов */
  // onChangeRowSelection?: OnChangeFn<RowSelectionState>
  // /** Обьект с состоянием чекбоксов */
  // selectionValue?: RowSelectionState
  // onFiltersChange?: (state: ColumnFiltersState) => void
  // /** Начальное состояние фильтров */
  // filtersState?: ColumnFiltersState
  // /**
  //  * Режим фильтрации
  //  * @default auto
  //  * auto - таблица сама фильтрует среди всех своих данных,
  //  * manual - фильтры не устанавливаются, а только вызывается событие onFiltersChange с состоянием фильтров
  //  */
  // // filterMode?: 'auto' | 'manual'
  // /** Включена ли группировка */
  // // enableGrouping?: boolean
}

const Table = ({
  columns = [],
  rows,
  className,
  maxLength = 5,
  // withRowSelection,
  withSorting,
  // withFiltering,
  // onChangeRowSelection,
  // selectionValue = {},
  // onFiltersChange,
  // filtersState = [],
  // enableGrouping = false,
  // filterMode = 'manual',
  // sortingMode = 'manual
  ...props
}: TableProps): ReactElement => {
  // const filterMode = 'manual'
  // const [rowSelection, setRowSelection] = useState(selectionValue)
  const [sorting, setSorting] = useState<SortingState>([])
  // const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>(
  //   filtersState || []
  // )
  // const [manualFilters, setManualFilters] = useState<ColumnFiltersState>(
  //   filtersState || []
  // )

  // const memoizedColumns = useMemo(() => {
  //   if (withRowSelection) {
  //     const deepCopy: Array<ColumnDef<any>> = JSON.parse(
  //       JSON.stringify(columns)
  //     )
  //     deepCopy.unshift({
  //       id: 'checkbox',
  //       header: ({ table }) => {
  //         return (
  //           <Checkbox
  //             checked={Boolean(table.getIsAllRowsSelected())}
  //             indeterminate={table.getIsSomeRowsSelected()}
  //             onChange={(value, e) =>
  //               table.getToggleAllRowsSelectedHandler().call({}, e)
  //             }
  //           />
  //         )
  //       },
  //       cell: ({ row }) => {
  //         return (
  //           <div className="px-1">
  //             <Checkbox
  //               checked={row.getIsSelected()}
  //               indeterminate={row.getIsSomeSelected()}
  //               onChange={row.getToggleSelectedHandler()}
  //             />
  //           </div>
  //         )
  //       }
  //     })
  //
  //     return deepCopy
  //   }
  //
  //   return columns
  // }, [...columns, withRowSelection])

  // const handleRowSelection: OnChangeFn<RowSelectionState> = (callback) => {
  //   const newState =
  //     typeof callback === 'function' ? callback(rowSelection) : {}
  //   setRowSelection(newState)
  //   onChangeRowSelection?.(newState)
  // }
  //
  // const handleFiltersChange: (value: string, column: Column<any>) => void = (
  //   value,
  //   column
  // ) => {
  // if (filterMode === 'auto') {
  //   column.setFilterValue(value)
  // }

  //   if (filterMode === 'manual') {
  //     const newState = [
  //       ...manualFilters?.filter((item) => item.id !== column.id),
  //       value ? { id: column.id, value } : undefined
  //     ].filter((item) => Boolean(item)) as ColumnFiltersState
  //     setManualFilters(newState)
  //     onFiltersChange?.(newState)
  //   }
  // }

  const table = useReactTable({
    data: rows,
    columns,
    getFilteredRowModel: getFilteredRowModel(),
    getCoreRowModel: getCoreRowModel(),
    state: {
      // rowSelection,
      // columnFilters,
      sorting
    },
    // filterFns: {
    //   elIncludesString: (row, columnId, filterValue) => {
    //     return getByText(row, columnId, filterValue)
    //   }
    // },
    // onColumnFiltersChange: setColumnFilters,
    onSortingChange: setSorting,
    // getSubRows: (row) => row?.subRows,
    // manualGrouping: enableGrouping,
    // onRowSelectionChange: handleRowSelection,
    getSortedRowModel: getSortedRowModel()
    // getFacetedRowModel: getFacetedRowModel(),
    // getFacetedUniqueValues: getFacetedUniqueValues(),
    // getFacetedMinMaxValues: getFacetedMinMaxValues()
  })

  const tableRows = table.getRowModel().rows.slice(0, maxLength)

  return (
    <table className={cn('inf-table', className)} {...props}>
      <TableHeader
        table={table}
        withSorting={withSorting}
        // withFiltering={withFiltering}
        // onFiltersChange={handleFiltersChange}
      />
      <TableBody
        rows={tableRows}
        // grouping={enableGrouping}
      />
    </table>
  )
}

export default Table
