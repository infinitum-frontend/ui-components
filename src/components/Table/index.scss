.inf-table {
  border-collapse: collapse;
  width: 100%;
  font-family: var(--inf-font-family-base);

  th {
    text-align: left;
    color: var(--inf-color-text-secondary);
    font-weight: var(--inf-font-weight-normal);
    font-size: var(--inf-text-size-small);
    padding: 6px;
  }

  tbody {
    tr {
      &:hover {
        background-color: var(--inf-control-background-color-hover);
      }
      &:first-child td {
        padding-top: 12px;
      }

      td {
        padding: 6px;
      }

      &.inf-table__group-label {
        background-color: #F2F4F5;
        text-transform: uppercase;

        td {
          padding: 0 6px;
        }
      }
    }
  }
}
