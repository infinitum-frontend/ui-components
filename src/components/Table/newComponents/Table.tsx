import { ComponentPropsWithoutRef, ReactElement } from 'react'
import TableBody from './TableBody'
import TableCell from './TableCell'
import TableFilterPopover from './TableFilterPopover'
import TableHeader from './TableHeader'
import TableHeaderCell from './TableHeaderCell'
import TableHeaderRow from './TableHeaderRow'
import TableRow from './TableRow'
import TableSort from './TableSort'

export interface TableProps extends ComponentPropsWithoutRef<'table'> {}

const Table = ({ children, ...restProps }: TableProps): ReactElement => {
  return (
    <table className="inf-new-table" {...restProps}>
      {children}
    </table>
  )
}

export default Object.assign(Table, {
  Header: TableHeader,
  HeaderCell: TableHeaderCell,
  HeaderRow: TableHeaderRow,
  Body: TableBody,
  Cell: TableCell,
  Row: TableRow,
  FilterPopover: TableFilterPopover,
  Sort: TableSort
})
