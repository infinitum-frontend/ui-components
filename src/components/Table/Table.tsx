// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, { ReactElement, useEffect, useMemo } from 'react'
import {
  Column,
  ColumnDef,
  ColumnMeta,
  ColumnResizeMode,
  getCoreRowModel,
  getFacetedUniqueValues,
  OnChangeFn as TanstackOnChangeFn,
  RowSelectionState,
  SortingState,
  useReactTable
} from '@tanstack/react-table'
import cn from 'classnames'
import TableHeader from './components/TableHeader'
import TableBody from './components/TableBody'
import { Checkbox } from 'Components/Checkbox'
import {
  TableColumnFiltersState,
  TableColumnFilterValue,
  TableRow,
  TableRowData,
  TableSelectedRow,
  TableSelectionState,
  TableVerticalAlignValue
} from './types'
import { OnChangeFn } from 'Utils/types'
import { mapRowToExternalFormat } from './helpers'
import TableBase, { TableBaseProps } from './components/TableBase'
import './Table.scss'

export interface TableProps extends TableBaseProps {
  /** Массив с данными для построения шапки таблицы */
  columns?: Array<ColumnDef<any, any>>
  /** Массив с данными для построения тела таблицы */
  rows?: Array<TableRowData<any>>
  /** Скругление границ таблицы */
  borderRadius?: 'xsmall' | 'small' | 'medium' | 'large'
  /** CSS свойство vertical-align для шапки */
  verticalAlignHead?: TableVerticalAlignValue
  /** CSS свойство vertical-align для рядов */
  verticalAlignBody?: TableVerticalAlignValue
  /**
   * @deprecated
   * Максимальное количество отображаемых элементов
   */
  maxLength?: number
  /** Включение сортировки по столбцам */
  withSorting?: boolean
  /** Начальное состояние сортировки */
  sortingState?: SortingState
  /** Событие изменения состояния сортировки */
  onSortingChange?: OnChangeFn<SortingState>
  // TODO: sorting mode auto
  /** @deprecated */
  withFiltering?: boolean
  /**
   * @deprecated
   * Событие изменения состояния фильтров
   */
  onFiltersChange?: OnChangeFn<TableColumnFiltersState>
  /**
   * @deprecated
   * Начальное состояние фильтров
   */
  filtersState?: TableColumnFiltersState
  /** Отображение чекбоксов в 1 колонке */
  withRowSelection?: boolean
  /** Событие изменения чекбоксов */
  onChangeRowSelection?: OnChangeFn<TableSelectionState<any>>
  /** Состояние выбора рядов через чекбокс */
  selectionState?: TableSelectionState<any>
  /**
   * Выбранный ряд. Если передается строка или число, идет сравнение аргумента с id ряда.
   */
  selectedRow?: TableSelectedRow
  onRowClick?: OnChangeFn<TableRow>
  /** Изменение ширины колонок
   * @value onChange изменение "вживую" при растягивании
   * @value onEnd изменение при отжатии кнопки мыши
   */
  resizeMode?: ColumnResizeMode
  /** видимость колонок
   * в качестве данных передается объект с ключами, соответствующими id колонок,
   * имеющим булевы значения, отражающими видимость колонки
   */
  columnVisibility?: Record<string, boolean>
  // /** Включена ли группировка */
  // // enableGrouping?: boolean
}

/** Компонент многофункциональной таблицы */
const Table = ({
  columns = [],
  rows = [],
  className,
  borderRadius,
  verticalAlignHead = 'top',
  verticalAlignBody,
  maxLength,
  withSorting,
  onSortingChange,
  sortingState = [],
  withFiltering,
  onFiltersChange,
  filtersState = [],
  withRowSelection,
  onChangeRowSelection,
  selectionState = [],
  selectedRow,
  onRowClick,
  resizeMode,
  columnVisibility = {},
  // enableGrouping = false,
  children,
  ...props
}: TableProps): ReactElement => {
  if (children) {
    return (
      <TableBase
        verticalAlignBody={verticalAlignBody}
        verticalAlignHead={verticalAlignHead}
        borderRadius={borderRadius}
        {...props}
      >
        {children}
      </TableBase>
    )
  }

  // Приводим состояние селекции столбцом к формату танстака(объект ключ-id ряда, значение-boolean)
  const tanstackSelectionState = useMemo(
    () =>
      selectionState?.reduce<RowSelectionState>((accumulator, currentValue) => {
        accumulator[currentValue.id] = true
        return accumulator
      }, {}),
    [selectionState]
  )

  // ==================== handlers ====================
  const handleSortingChange: (state: SortingState) => void = (state) => {
    onSortingChange?.(state)
  }

  const handleFiltersChange: (
    value: TableColumnFilterValue,
    filterType: ColumnMeta<any, any>['filterType'],
    column: Column<any>
  ) => void = (value, filterType, column) => {
    const newState = [
      ...filtersState?.filter((item) => item.id !== column.id),
      value ? { id: column.id, filterType, value } : undefined
    ].filter((item) => Boolean(item)) as TableColumnFiltersState
    onFiltersChange?.(newState)
  }

  // маппим данные ряда из формата танстака к нашему формату
  const handleRowSelection: TanstackOnChangeFn<RowSelectionState> = (
    callback
  ) => {
    const newTanstackSelectionState =
      typeof callback === 'function' ? callback(tanstackSelectionState) : {}

    const rowSelectionState: Array<TableRow<any>> = []
    Object.keys(newTanstackSelectionState).forEach((key) => {
      const row = mapRowToExternalFormat(table.getRow(key))
      rowSelectionState.push(row)
    })

    onChangeRowSelection?.(rowSelectionState)
  }

  const memoizedColumns = useMemo(() => {
    if (withRowSelection) {
      return [
        {
          id: 'checkbox',
          size: 20,
          header: ({ table }) => {
            return (
              <Checkbox
                checked={Boolean(table.getIsAllRowsSelected())}
                indeterminate={table.getIsSomeRowsSelected()}
                onClick={(e) => e.stopPropagation()}
                onChange={(value, e) =>
                  table.getToggleAllRowsSelectedHandler().call({}, e)
                }
              />
            )
          },
          cell: ({ row }) => {
            return (
              <div>
                <Checkbox
                  checked={row.getIsSelected()}
                  indeterminate={row.getIsSomeSelected()}
                  onClick={(e) => e.stopPropagation()}
                  onChange={row.getToggleSelectedHandler()}
                />
              </div>
            )
          }
        },
        ...columns
      ] as Array<ColumnDef<any>>
    }

    return columns
  }, [columns, withRowSelection])

  const table = useReactTable({
    data: rows,
    columns: memoizedColumns,
    columnResizeMode: resizeMode,
    getCoreRowModel: getCoreRowModel(),
    state: {
      rowSelection: tanstackSelectionState,
      sorting: sortingState,
      columnVisibility
    },
    manualFiltering: true,
    // manualGrouping: enableGrouping,
    onRowSelectionChange: handleRowSelection,
    getFacetedUniqueValues: getFacetedUniqueValues()
    // getSubRows: (row) => row?.subRows,
  })

  const tableRows = maxLength
    ? table.getRowModel().rows.slice(0, maxLength)
    : table.getRowModel().rows

  return (
    <table
      className={cn('inf-table', className, {
        [`inf-table--border-radius-${borderRadius as string}`]: borderRadius,
        'inf-table--bordered': props.bordered
      })}
      {...props}
    >
      <TableHeader
        table={table}
        withSorting={withSorting}
        withFiltering={withFiltering}
        sortingState={sortingState}
        onSortingChange={handleSortingChange}
        filtersState={filtersState}
        onFiltersChange={handleFiltersChange}
        resizeMode={resizeMode}
        verticalAlignHead={verticalAlignHead}
      />
      <TableBody
        rows={tableRows}
        selectedRow={selectedRow}
        onRowClick={onRowClick}
        // grouping={enableGrouping}
        verticalAlignBody={verticalAlignBody}
      />
    </table>
  )
}

export default Table
