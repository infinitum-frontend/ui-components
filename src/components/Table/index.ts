import {
  CellContext,
  createColumnHelper,
  createColumn,
  createCell,
  SortingState,
  ColumnDef,
  useReactTable,
  getCoreRowModel
} from '@tanstack/react-table'
import Table from './Table'
import { OnChangeFn } from 'Utils/types'
import { getNextSorting } from './helpers'

// TODO: стандартизировать список экспортов. Добавить нужные при необходимости
export * from './types'
export {
  Table,
  createColumnHelper as createTableColumnHelper,
  createCell as createTableCell,
  createColumn as createTableColumn,
  useReactTable as useTable,
  getCoreRowModel,
  getNextSorting
}
export type {
  OnChangeFn,
  CellContext as TableCellContext,
  SortingState as TableSortingState,
  ColumnDef as TableColumnDef
}
