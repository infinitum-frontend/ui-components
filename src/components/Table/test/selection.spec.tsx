// Тест кейсы:
// 1. наличие пропа withRowSelection отображает чекбоксы в 1 колонке +
// 2. наличие selectionState проставляет соответствующий чекбокс в статус Выбран +
// 3. нажатие на чекбокс у ряда вызывает функцию-колбэк с правильным состоянием +
// 4. выбор ряда проставляет в шапке чекбокс с состоянием indeterminate +
// 5. выбор чекбокса в шапке выбирает/сбрасывает все чекбоксы в таблице +
// 6. при динамическом изменении рядов(фильтрация/сортировка) состояния чекбоксов корректно привязываются к нужному ряду на основании getRowId

import { Portfolio, TABLE_COLUMNS, TABLE_DATA } from '../fixtures'
import { renderComponent } from '~/testSetup'
import Table from '../Table'
import { act, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TableSelectionState } from '../types'

describe('Table Row Selection', () => {
  it('should match snapshot', () => {
    const { el } = renderComponent(
      <Table rows={TABLE_DATA} columns={TABLE_COLUMNS} withRowSelection />
    )

    expect(el).toMatchSnapshot()
  })

  it('should render checkbox column if WithRowSelection', () => {
    renderComponent(
      <Table rows={TABLE_DATA} columns={TABLE_COLUMNS} withRowSelection />
    )

    const checkboxes = screen.queryAllByRole('checkbox')
    // Количество отрендеренных чекбоксов соответствует количеству рядов + 1 в шапке
    expect(checkboxes.length).toBe(TABLE_DATA.length + 1)

    // Все чекбоксы не выбраны
    checkboxes.forEach((checkbox) => {
      expect(checkbox).not.toBeChecked()
    })
  })

  it('should support default state', () => {
    const preselectedRowIndex = 2
    renderComponent(
      <Table
        rows={TABLE_DATA}
        columns={TABLE_COLUMNS}
        withRowSelection
        selectionState={[
          {
            id: String(preselectedRowIndex),
            rowData: TABLE_DATA[preselectedRowIndex]
          }
        ]}
      />
    )

    const checkboxes = screen.queryAllByRole('checkbox')

    // чекбокс в шапке имеет состояние indeterminate
    expect(checkboxes[0]).toBePartiallyChecked()
    // нужный чекбокс выбран
    expect(checkboxes[preselectedRowIndex + 1]).toBeChecked()

    checkboxes.forEach((checkbox, index) => {
      if (index !== 0 && index !== preselectedRowIndex + 1) {
        expect(checkbox).not.toBeChecked()
      }
    })
  })

  it('should emit correct state on change', async () => {
    const indexToCheck = 3
    let handlerPayload
    const mockChangeHandler = vi.fn((payload) => {
      handlerPayload = payload
    })
    const { rerender } = renderComponent(
      <Table
        rows={TABLE_DATA}
        columns={TABLE_COLUMNS}
        onChangeRowSelection={mockChangeHandler}
        withRowSelection
      />
    )

    const checkboxes = screen.queryAllByRole('checkbox')

    const user = userEvent.setup()
    await user.click(checkboxes[indexToCheck])

    expect(mockChangeHandler).toBeCalledTimes(1)
    expect(mockChangeHandler).toBeCalledWith(handlerPayload)

    rerender(
      <Table
        rows={TABLE_DATA}
        columns={TABLE_COLUMNS}
        onChangeRowSelection={mockChangeHandler}
        selectionState={handlerPayload}
        withRowSelection
      />
    )

    await user.click(checkboxes[indexToCheck])

    expect(mockChangeHandler).toBeCalledWith([])
  })

  it('should trigger all checkboxes on header checkbox change', async () => {
    let handlerPayload
    const mockChangeHandler = vi.fn((payload) => {
      handlerPayload = payload
    })

    const { rerender } = renderComponent(
      <Table
        rows={TABLE_DATA}
        columns={TABLE_COLUMNS}
        onChangeRowSelection={mockChangeHandler}
        withRowSelection
      />
    )

    const checkboxes = screen.queryAllByRole('checkbox')
    const user = userEvent.setup()
    await user.click(checkboxes[0])

    expect(mockChangeHandler).toBeCalledTimes(1)
    expect(mockChangeHandler).toBeCalledWith(handlerPayload)

    rerender(
      <Table
        rows={TABLE_DATA}
        columns={TABLE_COLUMNS}
        onChangeRowSelection={mockChangeHandler}
        selectionState={handlerPayload}
        withRowSelection
      />
    )

    checkboxes.forEach((checkbox) => {
      expect(checkbox).toBeChecked()
    })

    await user.click(checkboxes[0])

    expect(mockChangeHandler).toBeCalledWith([])
  })

  it('should keep correct state when rows change', () => {
    const preselectedRowIndex = 3

    const selectedElement = TABLE_DATA[preselectedRowIndex]
    const initialSelectionState: TableSelectionState<Portfolio> = [
      {
        id: selectedElement.portfolio,
        rowData: selectedElement
      }
    ]

    const { rerender } = renderComponent(
      <Table
        rows={TABLE_DATA}
        columns={TABLE_COLUMNS}
        getRowId={(row) => row.portfolio}
        selectionState={initialSelectionState}
        withRowSelection
      />
    )

    const checkboxes = screen.queryAllByRole('checkbox')

    expect(checkboxes[preselectedRowIndex + 1]).toBeChecked()

    act(() => {
      rerender(
        <Table
          rows={TABLE_DATA.filter((row) => {
            return row.portfolio.toLocaleLowerCase().match('машин')
          })}
          columns={TABLE_COLUMNS}
          getRowId={(row) => row.portfolio}
          selectionState={initialSelectionState}
          withRowSelection
        />
      )
    })

    const checkboxesAfterFiltering = screen.queryAllByRole('checkbox')

    // ожидаем, что после фильтрации выбранный чекбокс остался в таком же состоянии, т.к. благодаря getRowId селекция привязана не к индексу ряда, а к значению поля portfolio
    expect(checkboxesAfterFiltering[1]).toBeChecked()
  })
})
