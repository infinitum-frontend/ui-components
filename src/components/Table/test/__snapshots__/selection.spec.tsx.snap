// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Row Selection > should match snapshot 1`] = `
<table
  class="inf-table"
>
  <thead
    class="inf-table-header"
  >
    <tr
      class="inf-table-header-row"
    >
      <th
        class="inf-table-header-cell"
        colspan="1"
        style="width: 20px;"
      >
        <div
          class="inf-table-header-cell__content"
        >
          <div
            class="inf-table-header-cell__heading"
            data-selector="heading"
            data-tooltip-state="closed"
          >
            <label
              class="inf-checkbox"
            >
              <input
                aria-checked="false"
                aria-required="false"
                type="checkbox"
                value=""
              />
              <span
                class="inf-checkbox__box"
              >
                <svg
                  class="inf-checkbox__check-icon"
                  fill="none"
                  height="16px"
                  viewBox="0 0 20 20"
                  width="16px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6 9.70588L9.11111 13L13.5 6"
                    stroke="currentColor"
                    stroke-linejoin="round"
                    stroke-width="1.5"
                  />
                </svg>
              </span>
            </label>
          </div>
          <div
            class="inf-table-header-cell__buttons"
          />
        </div>
      </th>
      <th
        class="inf-table-header-cell"
        colspan="1"
        style="width: 150px;"
      >
        <div
          class="inf-table-header-cell__content"
        >
          <div
            class="inf-table-header-cell__heading"
            data-selector="heading"
            data-tooltip-state="closed"
          >
            Портфель
          </div>
          <div
            class="inf-table-header-cell__buttons"
          />
        </div>
      </th>
      <th
        class="inf-table-header-cell"
        colspan="1"
        style="width: 150px;"
      >
        <div
          class="inf-table-header-cell__content"
        >
          <div
            class="inf-table-header-cell__heading"
            data-selector="heading"
            data-tooltip-state="closed"
          >
            Показатель
          </div>
          <div
            class="inf-table-header-cell__buttons"
          />
        </div>
      </th>
      <th
        class="inf-table-header-cell"
        colspan="1"
        style="width: 150px;"
      >
        <div
          class="inf-table-header-cell__content"
        >
          <div
            class="inf-table-header-cell__heading"
            data-selector="heading"
            data-tooltip-state="closed"
          >
            Тип
          </div>
          <div
            class="inf-table-header-cell__buttons"
          />
        </div>
      </th>
      <th
        class="inf-table-header-cell"
        colspan="1"
        style="width: 150px;"
      >
        <div
          class="inf-table-header-cell__content"
        >
          <div
            class="inf-table-header-cell__heading"
            data-selector="heading"
            data-tooltip-state="closed"
          >
            Статус
          </div>
          <div
            class="inf-table-header-cell__buttons"
          />
        </div>
      </th>
      <th
        class="inf-table-header-cell"
        colspan="1"
        style="width: 150px;"
      >
        <div
          class="inf-table-header-cell__content"
        >
          <div
            class="inf-table-header-cell__heading"
            data-selector="heading"
            data-tooltip-state="closed"
          >
            Дата
          </div>
          <div
            class="inf-table-header-cell__buttons"
          />
        </div>
      </th>
    </tr>
  </thead>
  <tbody
    class="inf-table-body"
  >
    <tr
      class="inf-table-row inf-table-row--hoverable"
    >
      <td
        class="inf-table-cell"
      >
        <div>
          <label
            class="inf-checkbox"
          >
            <input
              aria-checked="false"
              aria-required="false"
              type="checkbox"
              value=""
            />
            <span
              class="inf-checkbox__box"
            >
              <svg
                class="inf-checkbox__check-icon"
                fill="none"
                height="16px"
                viewBox="0 0 20 20"
                width="16px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 9.70588L9.11111 13L13.5 6"
                  stroke="currentColor"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                />
              </svg>
            </span>
          </label>
        </div>
      </td>
      <td
        class="inf-table-cell"
      >
        ОПИФ Открытие
      </td>
      <td
        class="inf-table-cell"
      >
        2335
      </td>
      <td
        class="inf-table-cell"
      >
        УВНР
      </td>
      <td
        class="inf-table-cell"
      >
        Сформировано
      </td>
      <td
        class="inf-table-cell"
      >
        2024-10-31
      </td>
    </tr>
    <tr
      class="inf-table-row inf-table-row--hoverable"
    >
      <td
        class="inf-table-cell"
      >
        <div>
          <label
            class="inf-checkbox"
          >
            <input
              aria-checked="false"
              aria-required="false"
              type="checkbox"
              value=""
            />
            <span
              class="inf-checkbox__box"
            >
              <svg
                class="inf-checkbox__check-icon"
                fill="none"
                height="16px"
                viewBox="0 0 20 20"
                width="16px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 9.70588L9.11111 13L13.5 6"
                  stroke="currentColor"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                />
              </svg>
            </span>
          </label>
        </div>
      </td>
      <td
        class="inf-table-cell"
      >
        ЗПИФ И Открытие — ИММВБ — высокая капитализация ОПИФ И Открытие — ИММВБ — высокая капитализация ОПИФ И Открытие — ИММВБ — высокая капитализация ОПИФ И Открытие — ИММВБ — высокая капитализация
      </td>
      <td
        class="inf-table-cell"
      >
        2206
      </td>
      <td
        class="inf-table-cell"
      >
        УВУН
      </td>
      <td
        class="inf-table-cell"
      >
        Сформировано
      </td>
      <td
        class="inf-table-cell"
      />
    </tr>
    <tr
      class="inf-table-row inf-table-row--hoverable"
    >
      <td
        class="inf-table-cell"
      >
        <div>
          <label
            class="inf-checkbox"
          >
            <input
              aria-checked="false"
              aria-required="false"
              type="checkbox"
              value=""
            />
            <span
              class="inf-checkbox__box"
            >
              <svg
                class="inf-checkbox__check-icon"
                fill="none"
                height="16px"
                viewBox="0 0 20 20"
                width="16px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 9.70588L9.11111 13L13.5 6"
                  stroke="currentColor"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                />
              </svg>
            </span>
          </label>
        </div>
      </td>
      <td
        class="inf-table-cell"
      >
        Закрытие — Telecommunicaton Index
      </td>
      <td
        class="inf-table-cell"
      >
        2206
      </td>
      <td
        class="inf-table-cell"
      >
        УВУН
      </td>
      <td
        class="inf-table-cell"
      >
        Сформировано
      </td>
      <td
        class="inf-table-cell"
      >
        2020-01-01
      </td>
    </tr>
    <tr
      class="inf-table-row inf-table-row--hoverable"
    >
      <td
        class="inf-table-cell"
      >
        <div>
          <label
            class="inf-checkbox"
          >
            <input
              aria-checked="false"
              aria-required="false"
              type="checkbox"
              value=""
            />
            <span
              class="inf-checkbox__box"
            >
              <svg
                class="inf-checkbox__check-icon"
                fill="none"
                height="16px"
                viewBox="0 0 20 20"
                width="16px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 9.70588L9.11111 13L13.5 6"
                  stroke="currentColor"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                />
              </svg>
            </span>
          </label>
        </div>
      </td>
      <td
        class="inf-table-cell"
      >
        ЯРА И Открытие — ИММВБ — машиностроение
      </td>
      <td
        class="inf-table-cell"
      >
        2206
      </td>
      <td
        class="inf-table-cell"
      >
        УВНУ
      </td>
      <td
        class="inf-table-cell"
      >
        На согласовании
      </td>
      <td
        class="inf-table-cell"
      />
    </tr>
  </tbody>
</table>
`;
